parser/parser.go

```diff
@@ -904,3 +854,56 @@ func (p *parser) expandMacro(exprID int64, function string, target *exprpb.Expr,
 	}
 	return expr, true
 }
+
+func (p *parser) checkAndIncrementRecursionDepth() {
+	p.recursionDepth++
+	if p.recursionDepth > p.maxRecursionDepth {
+		panic(&recursionError{message: "max recursion depth exceeded"})
+	}
+}
+
+func (p *parser) decrementRecursionDepth() {
+	p.recursionDepth--
+}
+
+func unnest(tree antlr.ParseTree) antlr.ParseTree {
+	for {
+		switch t := tree.(type) {
+		case *gen.ExprContext:
+			if t.GetOp() != nil {
+				return t
+			}
+			tree = t.GetE()
+		case *gen.ConditionalOrContext:
+			if t.GetOps() != nil && len(t.GetOps()) > 0 {
+				return t
+			}
+			tree = t.GetE()
+		case *gen.ConditionalAndContext:
+			if t.GetOps() != nil && len(t.GetOps()) > 0 {
+				return t
+			}
+			tree = t.GetE()
+		case *gen.RelationContext:
+			if t.GetOp() != nil {
+				return t
+			}
+			tree = t.Calc()
+		case *gen.CalcContext:
+			if t.GetOp() != nil {
+				return t
+			}
+			tree = t.Unary()
```

Comment:
Are we ok not checking for the existence of unary here? Especially since we're removing the nil checks from the VisitCalc method. The same question applies for RelationContext above

Reply:
I should probably add a condition in the for-loop that tests for `tree != nil` since that's the worst case scenario for traversing an incomplete parse graph that results from a syntax error.